### ISSUE TITLE
Service Account token expiring every few days

### ISSUE THREAD

**Issue by nikita-tkachov** [1Password/onepassword-sdk-js#106]
### Scenario & Reproduction Steps

Service account token recently started erroring out with the error `error when retrieving vault metadata: http error: unexpected http status: 401 Unauthorized` every couple of days.

The fix for the issue has been to regenerate the token for the Service Account.

I don't have exact reproduction steps right now but it only started happening recently. First occurrence from Sentry logs is `Oct 25, 2024 9:11:29 PM UTC`

### Actual Behavior

Service account token expires every few days.

### Expected Behavior

When I set the token to "never expire", it should mean that it never expires!

### SDK version

0.3.0

**Comment by hculea** [comment]
Hey @nikita-tkachov, 

Thank you for reporting! This is indeed unexpected. I've got a couple of questions on this:
1. Is the SDK version that you are using 0.1.3?
2. Is this issue consistently happening on some vaults, or sporadically on operations that usually work?
3. Have you rotated your service account you use or made changes to the vaults on which this error is triggered, recently? 

To identify whether this is indeed an issue with the service account or with the SDK, we can try running a command with the [1Password CLI](https://developer.1password.com/docs/cli/get-started/). 

Exporting the service account like `export OP_SERVICE_ACCOUNT_TOKEN=<your-service-account-token>` and running `op vault get <the-vault-with-an-issue>` will help us determine that. Also, does `op vault list` contain the output that you would expect (i.e. all the vaults you gave permissions for)?

Looking forward to hearing from you! Thanks,

**Comment by hculea** [comment]
@nikita-tkachov Thanks for the details!

I'd be interested in the order of magnitude of the number of commands that you run (tens? hundreds? thousands?). My main hypothesis right now is that this may be a rate-limiting issue with a wrong error text and status code. If this is the case, the service account tokens would actually re-begin working later on.

You can read more about this here: https://developer.1password.com/docs/service-accounts/rate-limits/ 

However, this doesn't seem to correlate with the service account working with the CLI. You can check yourself if this is the case by running `op service-account ratelimit <service-account-name>` in the CLI, on a service account that appears to be "expired".

But I can look up for more details knowing your account. Can you please send your account name / domain (.com/.ca/.eu) / UUID to support@1password.com, referencing this GitHub thread? I'll be able to pick it up from there and look for more information in the server logs.

**Comment by hculea** [comment]
Thanks for helping with debugging this, @nikita-tkachov ! I can also confirm we've received your message on the support email.

> 
It appears the issue may have been on my side. We were injecting createClient as a dependency into our service methods and re-using it for all API calls.

We've switched to creating the client on every single API call. This significantly slows things down (3-5 seconds for simple GET operations to grab the password), but completely fixes the issue.

> I assume it's something with the SDK where a short-lived token is generated when the client is created, but not refreshed later on, causing the 401 errors.

That should not be the case - while the duration of a service account session is 30 minutes by default, this should be refreshed automatically under the hood, which would point at a bug with the said session refresher.

I'm going to open up a ticket about this internally and bring it up with our team for the next sprint planning. Is the workaround that you're currently employing unblocking you, for the time being?

**Comment by MOmarMiraj** [comment]
Hi @haaksmash and @nikita-tkachov 

With the latest version of the JS SDK, this bug has been resolved! Feel free to open another issue if you encounter any more bugs concerning this topic.
